File: app/src/main/java/com/zaze/demo/component/device/DeviceViewModel.java
package com.zaze.demo.component.device;

import android.app.Application;
import android.view.Display;

import androidx.lifecycle.MutableLiveData;

import com.zaze.common.base.AbsAndroidViewModel;
import com.zaze.demo.model.DeviceModel;
import com.zaze.demo.model.ModelFactory;
import com.zaze.demo.model.entity.DeviceStatus;
import com.zaze.utils.DisplayUtil;
import com.zaze.utils.NetUtil;
import com.zaze.utils.ZStringUtil;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

/**
 * Description :
 *
 * @author : zaze
 * @version : 2017-01-22 01:39 1.0
 */
public class DeviceViewModel extends AbsAndroidViewModel {
    private DeviceModel deviceModel;
    MutableLiveData<ArrayList<DeviceStatus>> deviceInfoList = new MutableLiveData<>();
    MutableLiveData<String> inchData = new MutableLiveData<>();


    public DeviceViewModel(@NotNull Application application) {
        super(application);
        deviceModel = ModelFactory.getDeviceModel();
    }

    public void loadDeviceInfo() {
        deviceInfoList.setValue(deviceModel.getDeviceInfo());
    }

    /**
     * 计算物理尺寸
     *
     * @param dpi dpi
     */
    public void calculatePhysicalSize(int dpi) {
        if (dpi < 1) {
            dpi = DisplayUtil.getMetrics().densityDpi;
        }
        double inch = Math.sqrt(Math.pow(DisplayUtil.getMetrics().widthPixels, 2) + Math.pow(DisplayUtil.getMetrics().heightPixels, 2)) / dpi;
        inchData.setValue(ZStringUtil.format("%.2f寸", inch));
    }

    public void calculateDpi(String inch) {
        try {
            float inchNum = Float.parseFloat(inch);
            if (inchNum > 0) {
                double dpi = Math.sqrt(Math.pow(DisplayUtil.getMetrics().widthPixels, 2) + Math.pow(DisplayUtil.getMetrics().heightPixels, 2)) / inchNum;
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/com/zaze/demo/component/device/DeviceViewModel.kt
package com.zaze.demo.component.device

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.zaze.common.base.AbsAndroidViewModel
import com.zaze.common.base.AbsViewModel
import com.zaze.common.base.BaseApplication
import com.zaze.core.data.repository.DeviceRepository
import com.zaze.core.model.data.DeviceStatus
import com.zaze.utils.DisplayUtil
import com.zaze.utils.ZStringUtil
import dagger.hilt.android.AndroidEntryPoint
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.*
import javax.inject.Inject

/**
 * Description :
 *
 * @author : zaze
 * @version : 2017-01-22 01:39 1.0
 */
@HiltViewModel
class DeviceViewModel @Inject constructor(
    application: Application
) : AbsAndroidViewModel(application) {

    @Inject
    lateinit var deviceModel: DeviceRepository

    val deviceInfoList = MutableLiveData<List<DeviceStatus>>()
    val inchData = MutableLiveData<String>()

    fun loadDeviceInfo() {
        viewModelScope.launch(Dispatchers.IO) {
            deviceInfoList.postValue(deviceModel.getDeviceInfo())
        }
    }

    /**
     * 计算物理尺寸
     *
     * @param dpi dpi
     */
    fun calculatePhysicalSize(iDpi: Int) {
        val dpi = if (iDpi < 1) {
            DisplayUtil.getMetrics().densityDpi
        } else {
            iDpi
        }
        val displayProfile = DisplayUtil.getDisplayProfile()
//        displayProfile.realWidthPixels}x${displayProfile.realHeightPixels
        val inch = Math.sqrt(
            Math.pow(displayProfile.realWidthPixels.toDouble(), 2.0) + Math.pow(
                displayProfile.realHeightPixels.toDouble(), 2.0
            )
        ) / dpi
        inchData.value = ZStringUtil.format("%.2f寸", inch)
    }

    fun calculateDpi(inch: String) {
        try {
            val inchNum = inch.toFloat()
            if (inchNum > 0) {
                val dpi = Math.sqrt(
                    Math.pow(DisplayUtil.getMetrics().widthPixels.toDouble(), 2.0) + Math.pow(
                        DisplayUtil.getMetrics().heightPixels.toDouble(), 2.0
                    )
                ) / inchNum
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}